### ISSUE TITLE
Crash when card text is too long

### ISSUE THREAD

**Issue by Swarmotte** [chilli-axe/mtg-photoshop-automation#32]
HI,
Photoshop crashes for some cards like God-Eternal Kefnet or Lich's Mastery which have long rule text. I tried to find out how to correct this error but I can't.

**Comment by MrTeferi** [comment]
I was getting this same issue with rendering Toxrill but only with certain templates?

**Comment by MrTeferi** [comment]
2 more cards that crash photoshop when you try to render them:
Ao, the Dawn Sky
Reckoner Bankbuster

In keeping with the trend these both have a lot of text. I'm really not sure what could be causing it, would love to hear ideas if anyone might know?

**Comment by MrTeferi** [comment]
I'm doing some testing with Ao, the Dawn Sky, this problem does indeed have to do with the text contents, specifically it seems for CreatureFormattedTextArea. I have yet to find a noncreature with long text that crashes the app. I took these lines:
```
new CreatureFormattedTextArea(
                    layer = rules_text,
                    text_contents = this.layout.oracle_text,
                    text_colour = get_text_layer_colour(rules_text),
                    flavour_text = this.layout.flavour_text,
                    is_centred = is_centred,
                    reference_layer = text_and_icons.layers.getByName(LayerNames.TEXTBOX_REFERENCE),
                    pt_reference_layer = text_and_icons.layers.getByName(LayerNames.PT_REFERENCE),
                    pt_top_reference_layer = text_and_icons.layers.getByName(LayerNames.PT_TOP_REFERENCE),
                ),
```
And I changed text_contents to a random small string like "eedfwefwefg" and the card renders perfectly with that text. So something about Ao, the Dawn Sky's text is causing the crash

**Comment by ndepaola** [comment]
turns out that photoshop crashes when you try to insert text into a textbox that's not large enough to contain it:

![image](https://user-images.githubusercontent.com/3079166/154254211-4b386e9e-e3ab-4840-9e3a-b98d2207590a.png)

increasing the height of the rules text areas will solve this. in the meantime i'll have a think about how to circumvent this issue without resorting to template updates. curious that i couldn't recreate this issue with kefnet but i could with ao

**Comment by ndepaola** [comment]
greater morphling works fine so it's not a character limit

![image](https://user-images.githubusercontent.com/3079166/154257145-ac0522ec-6931-43f8-a3f7-2c55300ff804.png)

**Comment by MrTeferi** [comment]
Yeah neither Kefnet nor Lich's Mastery were crashing on my end either, but that is an interesting find regarding the text box size!

**Comment by MrTeferi** [comment]
So I think I have a better way to fix this issue than the ideas I brainstormed in the past, here's what I did for proxyshop that seems to work reliably:
* If FormattedTextArea contents has over 280 characters, make duplicate of noncreature rules text (I actually test if the noncreature rules text layer exists and if it doesn't use just rules text, i.e. classic template, I think this is possible in javascript?)
* Set the deplicate TextItem height to 2000 and insert the card's text including flavor text
* scale_text_to_fit_height() -- New function similar to scale to fit reference, feed it the height of the reference increased by 1%, then scale down the font size and leading till it fits in this bound
* set your intended TextItem to the newly scaled font size/leading
* remove the duplicate TextItem
* everything else in the text_layers class runs as normal afterwards (insert text, format it, scale it, add divider, etc)

The main reason I scale it based on the height of the reference+1% is to account for size adjustment that may happen after symbols etc get formatted, then the original scale_text_to_fit_reference takes care of any excess!
